src/backend/distributed/commands/role.c

```diff
@@ -141,6 +245,100 @@ ExtractEncryptedPassword(Oid roleOid)
 }
 
 
+/*
+ * GenerateAlterRoleSetIfExistsCommandList generate a list of ALTER ROLE .. SET commands that
+ * copies a role session defaults from the pg_db_role_settings table.
+ */
+static List *
+GenerateAlterRoleSetIfExistsCommandList(HeapTuple tuple, TupleDesc
+										DbRoleSettingDescription)
+{
+	AlterRoleSetStmt *stmt = makeNode(AlterRoleSetStmt);
+	const char *currentDatabaseName = CurrentDatabaseName();
+	List *commandList = NIL;
+	bool isnull = false;
+
+	const char *databaseName =
+		GetDatabaseNameFromDbRoleSetting(tuple, DbRoleSettingDescription);
+
+	/*
+	 * session defaults for databases other than the current one are skipped
+	 */
+	if (databaseName != NULL &&
+		pg_strcasecmp(databaseName, currentDatabaseName) != 0)
+	{
+		ereport(NOTICE, (errmsg("Citus partially supports ALTER ROLE .. IN DATABASE"
+								" .. SET  for distributed databases"),
+						 errdetail("Citus propagates the session defaults that affect "
+								   "current database"),
+						 errhint("You can manually change attributes of roles "
+								 "on workers")));
+		return NULL;
+	}
+
+	if (databaseName != NULL)
+	{
+		stmt->database = pstrdup(databaseName);
+	}
+
+	const char *roleName = GetRoleNameFromDbRoleSetting(tuple, DbRoleSettingDescription);
+
+	/*
+	 * default roles are skipped, because reserved roles
+	 * cannot be altered.
+	 */
+	if (roleName != NULL && IsReservedName(roleName))
+	{
+		return NULL;
+	}
+
+	if (roleName != NULL)
+	{
+		stmt->role = makeNode(RoleSpec);
+		stmt->role->location = -1;
+		stmt->role->roletype = ROLESPEC_CSTRING;
+		stmt->role->rolename = pstrdup(roleName);
+	}
+
+	Datum setconfig = heap_getattr(tuple, Anum_pg_db_role_setting_setconfig,
+								   DbRoleSettingDescription, &isnull);
+
+	Datum *configs;
+	int nconfigs;
+	int i;
+
+	deconstruct_array(DatumGetArrayTypeP(setconfig),
+					  TEXTOID, -1, false, 'i',
+					  &configs, NULL, &nconfigs);
+
+	for (i = 0; i < nconfigs; i++)
+	{
+		char *config = TextDatumGetCString(configs[i]);
+
+		char *seperator = strchr(config, '=');
+
+		/*
+		 * Each array element should have the form name=value.  If the "="
+		 * is missing for some reason, ignore it
+		 */
+		if (seperator == NULL)
+		{
+			continue;
+		}
+
+		*seperator++ = '\0';
```

Comment:
instead might be better to use [`ParseLongOption `](https://github.com/postgres/postgres/blob/3d475515a15f70a4a3f36fbbba93db6877ff8346/src/backend/utils/misc/guc.c#L10413)

Reply:
Introduced `ParseConfigOption` that behaves slightly differently compared to `ParseLongOption`

```c
/*
 * ParseConfigOption takes a string of the form "some-option=some value" and
 * returns name = "some-option" and value = "some value" in malloc'ed
 * storage.
 *
 * Most of the logic is copied from ParseLongOption in backend/utils/misc/guc.c
 * with one major difference: '-' is not converted to '_' in the option name here.
 *
 * If there is no '=' in the input string then value will be NULL.
 */
static void
ParseConfigOption(const char *string, char **name, char **value)
```

Reply:
hmm funky for the `-` => `_` conversion in `ParseLongOption`. The reason I was suggesting to use that function is because that is the exact function Postgres uses to parse the options from the catalog table on backend startup.

I will have a look into the significance of `-` in options names.